File: close_all_positions.py
#!/usr/bin/env python3
import requests
import time
import hmac
import hashlib
import json

# Bybit Demo API credentials
API_KEY = "lCMnwPKIzXASNWn6UE"
API_SECRET = "aXjs1SF9tmW3riHMktmjtyOyAT85puvrVstr"
BASE_URL = "https://api-demo.bybit.com"

def generate_signature(timestamp, params_str):
    """Generate signature for Bybit API"""
    string_to_sign = f"{timestamp}{API_KEY}5000{params_str}"
    return hmac.new(
        API_SECRET.encode('utf-8'),
        string_to_sign.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()

def get_positions():
    """Get all positions from Bybit"""
    endpoint = "/v5/position/list"
    timestamp = str(int(time.time() * 1000))
    
    params = "category=linear&settleCoin=USDT"
    signature = generate_signature(timestamp, params)
    
    headers = {
        "X-BAPI-API-KEY": API_KEY,
        "X-BAPI-TIMESTAMP": timestamp,
        "X-BAPI-RECV-WINDOW": "5000",
        "X-BAPI-SIGN": signature
    }
    
    url = f"{BASE_URL}{endpoint}?{params}"
    
    try:
        response = requests.get(url, headers=headers)
        data = response.json()
        
        if data.get("retCode") == 0:
            positions = data.get("result", {}).get("list", [])
            return [pos for pos in positions if float(pos.get("size", "0")) > 0]
        else:
            print(f"❌ Error getting positions: {data.get('retMsg', 'Unknown error')}")
            return []
            
    except Exception as e:
        print(f"❌ Exception getting positions: {e}")
        return []

def close_position(symbol, side, size):
    """Close a position by placing opposite order"""
    endpoint = "/v5/order/create"
    timestamp = str(int(time.time() * 1000))
    
    # Determine opposite side
    close_side = "Sell" if side == "Buy" else "Buy"
    
    # Create order data
    order_data = {
        "category": "linear",
        "symbol": symbol,
        "side": close_side,
        "orderType": "Market",
        "qty": str(size),
        "timeInForce": "GTC",
        "positionIdx": 0
    }
    
    json_string = json.dumps(order_data, separators=(',', ':'))
    signature = generate_signature(timestamp, json_string)
    
    headers = {
        "X-BAPI-API-KEY": API_KEY,
        "X-BAPI-TIMESTAMP": timestamp,
        "X-BAPI-RECV-WINDOW": "5000",
        "X-BAPI-SIGN": signature,
        "Content-Type": "application/json"
    }
    
    url = f"{BASE_URL}{endpoint}"
    
    try:
        response = requests.post(url, headers=headers, data=json_string)
        data = response.json()
        
        if data.get("retCode") == 0:
            order_id = data.get("result", {}).get("orderId", "")
            print(f"✅ Closed {symbol} position: Order ID {order_id}")
            return True
        else:
            print(f"❌ Failed to close {symbol}: {data.get('retMsg', 'Unknown error')}")
            return False
            
    except Exception as e:
        print(f"❌ Exception closing {symbol}: {e}")
        return False

def main():
    print("🔄 Getting all positions...")
    positions = get_positions()
    
    if not positions:
        print("✅ No positions to close!")
        return
    
    print(f"📊 Found {len(positions)} positions to close:")
    
    for pos in positions:
        symbol = pos.get("symbol", "")
        side = pos.get("side", "")
        size = float(pos.get("size", "0"))
        notional = size * float(pos.get("markPrice", "0"))
        
        print(f"   {symbol}: {side} {size} (${notional:,.2f})")
    
    print("\n🚨 WARNING: This will close ALL positions!")
    confirm = input("Type 'YES' to confirm: ")
    
    if confirm != "YES":
        print("❌ Cancelled")
        return
    
    print("\n🔄 Closing all positions...")
    
    success_count = 0
    for pos in positions:
        symbol = pos.get("symbol", "")
        side = pos.get("side", "")
        size = float(pos.get("size", "0"))
        
        if close_position(symbol, side, size):
            success_count += 1
        
        time.sleep(0.1)  # Rate limiting
    
    print(f"\n✅ Successfully closed {success_count}/{len(positions)} positions")
    print("🎯 All positions closed! You can now restart the trading system with clean slate.")

if __name__ == "__main__":
    main()

File: close-all-positions.js
/**
 * Close All Open Positions on Bybit Demo
 * This will free up capital for the 12 USDT trading system
 */

const bybitClient = require('./ui/dashboard-backend/src/utils/bybit-client');
const logger = require('./ui/dashboard-backend/src/utils/logger');

async function closeAllPositions() {
  try {
    console.log('🔍 Getting all open positions...');
    
    // Get all positions
    const positions = await bybitClient.getPositions();
    
    if (!positions || positions.retCode !== 0) {
      throw new Error('Failed to get positions');
    }
    
    const openPositions = positions.result.list.filter(pos => parseFloat(pos.size) > 0);
    
    console.log(`📊 Found ${openPositions.length} open positions to close`);
    
    if (openPositions.length === 0) {
      console.log('✅ No open positions to close');
      return;
    }
    
    // Close each position
    let closedCount = 0;
    let failedCount = 0;
    
    for (const position of openPositions) {
      try {
        console.log(`🔄 Closing ${position.symbol} ${position.side} position (Size: ${position.size})`);
        
        // Determine opposite side to close position
        const closeSide = position.side === 'Buy' ? 'Sell' : 'Buy';
        
        // Place market order to close position
        const closeOrder = {
          category: 'linear',
          symbol: position.symbol,
          side: closeSide,
          orderType: 'Market',
          qty: position.size,
          timeInForce: 'IOC',
          reduceOnly: true  // Important: this closes the position
        };
        
        const response = await bybitClient.placeOrder(closeOrder);
        
        if (response && response.retCode === 0) {
          console.log(`✅ Closed ${position.symbol}: Order ID ${response.result.orderId}`);
          closedCount++;
        } else {
          console.log(`❌ Failed to close ${position.symbol}: ${response?.retMsg}`);
          failedCount++;
        }
        
        // Small delay to avoid rate limits
        await new Promise(resolve => setTimeout(resolve, 500));
        
      } catch (error) {
        console.error(`❌ Error closing ${position.symbol}:`, error.message);
        failedCount++;
      }
    }
    
    console.log('');
    console.log('📊 POSITION CLOSURE SUMMARY:');
    console.log(`   ✅ Successfully closed: ${closedCount} positions`);
    console.log(`   ❌ Failed to close: ${failedCount} positions`);
    console.log(`   📈 Total processed: ${openPositions.length} positions`);
    
    // Wait a moment then check final balance
    console.log('');
    console.log('⏳ Waiting 5 seconds for positions to settle...');
    await new Promise(resolve => setTimeout(resolve, 5000));
    
    // Check final balance
    const finalBalance = await bybitClient.getWalletBalance({
      accountType: 'UNIFIED'
    });
    
    if (finalBalance && finalBalance.retCode === 0) {
      const account = finalBalance.result.list[0];
      console.log('💰 FINAL ACCOUNT STATUS:');
      console.log(`   Total Available Balance: ${account.totalAvailableBalance} USD`);
      console.log(`   Total Initial Margin: ${account.totalInitialMargin} USD`);
      
      const availableForTrading = parseFloat(account.totalAvailableBalance);
      console.log(`   🎯 Available for 12 USDT Trading: ${availableForTrading.toFixed(2)} USD`);
      
      if (availableForTrading >= 12) {
        console.log('   ✅ Ready for 12 USDT trading system!');
      } else {
        console.log('   ⚠️ May need to close more positions or wait for settlement');
      }
    }
    
  } catch (error) {
    console.error('❌ Error in closeAllPositions:', error);
  }
}

closeAllPositions();
